File: java/src/1002.java
import static org.assertj.core.api.Assertions.assertThat;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<String> commonChars(String[] words) {
        int[] freq = new int[26];
        for (int i = 0, n = words[0].length(); i < n; i++) {
            freq[words[0].charAt(i) - 'a']++;
        }
        for (int i = 1, n = words.length; i < n; i++) {
            int[] temp = new int[26];
            for (int j = 0, m = words[i].length(); j < m; j++) {
                temp[words[i].charAt(j) - 'a']++;
            }
            for (int k = 0; k < 26; k++) {
                freq[k] = Math.min(freq[k], temp[k]);
            }
        }
        var chars = new ArrayList<String>();
        for (int i = 0; i < 26; i++) {
            while (freq[i]-- > 0) {
                chars.add(String.valueOf((char) (i + 'a')));
            }
        }
        return chars;
    }

    public static void main(String[] args) {
        var o = new Solution();

        assertThat(o.commonChars(new String[] { "bella", "label", "roller" })).isEqualTo(List.of("e", "l", "l"));
        assertThat(o.commonChars(new String[] { "cool", "lock", "cook" })).isEqualTo(List.of("c", "o"));

        var start = Instant.now();
        for (var i = 0; i < 1000; i++) {
            o.commonChars(new String[] { "bella", "label", "roller" });
            o.commonChars(new String[] { "cool", "lock", "cook" });
        }
        System.out.println(ChronoUnit.MICROS.between(start, Instant.now()));
    }
}

File: cpp/src/1002.cpp
/*
[1002. Find Common Characters (Easy)](https://leetcode.com/problems/find-common-characters/)

@topic: Array
@topic: Hash Table
@topic: String

Constraints:
  - `1 <= words.length <= 100`
  - `1 <= words[i].length <= 100`
  - `words[i]` consists of lowercase English letters.
*/

#include <algorithm>
#include <string>
#include <vector>

using std::min;
using std::string;
using std::vector;

class Solution {
   public:
    vector<string> commonChars(vector<string>& words) {
        int freq[26]{};
        for (char c : words[0]) ++freq[c - 'a'];
        for (int i = 1, n = words.size(); i < n; ++i) {
            int hash[26]{};
            for (char c : words[i]) ++hash[c - 'a'];
            for (int i = 0; i < 26; ++i) freq[i] = min(freq[i], hash[i]);
        }
        vector<string> chars;
        for (int i = 0; i < 26; ++i)
            while (freq[i]--) chars.emplace_back(1, i + 'a');
        return chars;
    }
};

#include <cassert>
#include <chrono>
#include <iostream>

using std::tuple;
using std::chrono::duration_cast;
using std::chrono::microseconds;
using std::chrono::system_clock;

int main() {
    Solution o;

    vector<tuple<vector<string>, vector<string>>> CASES = {
        {{"bella", "label", "roller"}, {"e", "l", "l"}},
        {{"cool", "lock", "cook"}, {"c", "o"}},
    };

    for (auto& [words, excepted] : CASES) {
        assert(o.commonChars(words) == excepted);
    }

    auto start = system_clock::now();
    for (auto& [words, _] : CASES) {
        for (int i = 0; i < 100000; ++i) {
            o.commonChars(words);
        }
    }
    auto end = system_clock::now();
    std::cout << duration_cast<microseconds>(end - start).count() << std::endl;
}
